### PR TITLE
Tomm/detox update

### PR DESCRIPTION
**Pull request by MuckT**
### Description

The goal of this PR is to reduce test flake within our CICD pipeline, to setup e2e test parallelization and add additional screen comparison utilities.

This PR also includes an update to the Detox Framework  from `17.13.2` to `18.14.1`

Refactored / Added Tests:
| Test Suite | Use cases |
| ----- | ----- |
| Account Setup | NewAccountOnboarding |
| Account Setup | RestoreAccountOnboarding |
| AccountManagement | Support |
| AccountManagement | Settings |
| AccountManagement | Reset Account |
| Ramps | Cashin |

### Other changes

Changes to test utilities: `enterPinUiIfNecessary`, `enterPinUi` and `isElementVisible`

Added to test utilities: `pixelDiff`, `setDemoMode`, `scrollIntoView`, `quickOnboarding` and `isElementExistent`

Updated `e2e/README.md` to reflect e2e test framework updates.

Android Command Line tools were upgraded.

### Tested

Ran locally on Android and iOS emulators - Circle Ci Results Passing

Run unit tests - Circle Ci Results Passing

### Backwards compatibility

This should only impact e2e tests. Snapshots have been updated for mobile tests.

### CODE REVIEW

File: packages/mobile/e2e/src/usecases/Support.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import { pixelDiff, scrollIntoView } from '../utils/utils'
+
+export default Support = () => {
+  beforeEach(async () => {
+    await device.reloadReactNative()
+    await dismissBanners()
+  })
+
+  if (device.getPlatform() === 'ios') {
+    it("Display 'Contact' on Shake", async () => {
```
Comment (silasbw) [inline packages/mobile/e2e/src/usecases/Support.js:11]: mixed used of `"` and `'` -- i'm surprised a linter / styler doesn't catch this.  Should it?

File: packages/mobile/e2e/src/usecases/SecureSend.js [hunk 2]
```diff
@@ -1,11 +1,19 @@
 import { enterPinUiIfNecessary, inputNumberKeypad, sleep } from '../utils/utils'
+import { dismissBanners } from '../utils/banners'
+var faker = require('faker')
```
Comment (silasbw) [inline packages/mobile/e2e/src/usecases/SecureSend.js:3]: FWIW, in general eschew `var` usage in favor of `let` or `const`.

For module imports, it would be good to converge on using `imports` (or the `import faker = require('faker')` depending what faker exports).

File: packages/mobile/e2e/src/usecases/RestoreAccountOnboarding.js [hunk 3]
```diff
@@ -1,16 +1,24 @@
-import { enterPinUi, sleep, waitForElementId } from '../utils/utils'
-import { SAMPLE_BACKUP_KEY } from '../utils/consts'
-
-const EXAMPLE_NAME = 'Test Name'
+import { enterPinUi, waitForElementId, sleep } from '../utils/utils'
```
Comment (silasbw) [inline packages/mobile/e2e/src/usecases/RestoreAccountOnboarding.js:1]: Is this a self-contained set of changes to specific test cases?  If so, it's a candidate for splitting into a separate PR to facilitate review.

File: packages/mobile/e2e/src/usecases/Send.js [hunk 4]
```diff
@@ -18,8 +26,10 @@ export default Send = () => {
     await inputNumberKeypad(AMOUNT_TO_SEND)
     await element(by.id('Review')).tap()
 
+    await sleep(1000)
```
Comment (silasbw) [inline packages/mobile/e2e/src/usecases/Send.js:29]: I see comments elsewhere around "if we aren't sleeping long enough, this test will fail".

Is there a way to avoid `sleep` in this case and instead wait for some explicit state (with a timeout)?  E.g., wait for a specific element to appear or change state.

File: packages/mobile/e2e/src/usecases/Settings.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { dismissBanners } from '../utils/banners'
```
Comment (silasbw) [inline packages/mobile/e2e/src/usecases/Settings.js:1]: Since this is wholly new it seems like a candidate for a separate PR.

File: packages/mobile/e2e/src/utils/utils.js [hunk 6]
```diff
@@ -95,3 +132,121 @@ export function quote(s) {
   // while on android it does not, so we add it
   return device.getPlatform() === 'ios' ? s : `"${s}"`
 }
+
+export async function quickOnboarding() {
```
Comment (silasbw) [inline packages/mobile/e2e/src/utils/utils.js:136]: Since these are brand new tests, seems like a candidate for a separate PR.

**Review by silasbw — COMMENTED** [review]
Random uneducated feedback from a novice.

File: packages/mobile/e2e/src/usecases/Support.js (hunk 1 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/Support.js:11]: Ideally it would be `'Display \'Contact\' on Shake'`, but VScode reformats to above regardless of my VScode format on save options.

File: packages/mobile/e2e/src/usecases/Settings.js (hunk 5 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/Settings.js:1]: In the future, I'd like to have additional use cases in their own PR; however, as the tests are restructured I wanted to avoid submitting a PR which would break existing tests if merged.

File: packages/mobile/e2e/src/usecases/SecureSend.js (hunk 2 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/SecureSend.js]: - [x] Fixed use of var in commit [458cd4f](https://github.com/celo-org/wallet/pull/260/commits/458cd4fdcb98574a3f3591e5f372d1e1a095fbe6)

Module imports would be good to have clarification / standards on.

File: packages/mobile/e2e/src/usecases/Send.js (hunk 4 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/Send.js:29]: Locally the tests will pass without the sleep, but in running in Circle Ci the sleeps are needed to reduce test flake. This can be reassessed when setting up the new CICD pipeline.

File: .circleci/config.yml [hunk 7]
```diff
@@ -187,8 +187,8 @@ jobs:
           name: install android tools
           command: |
             sh packages/mobile/scripts/install-android-tools.sh
-            echo 'export PATH=$ANDROID_SDK_ROOT/tools/bin:$PATH'  >> $BASH_ENV
-            echo 'export PATH=$ANDROID_SDK_ROOT/tools:$PATH'  >> $BASH_ENV
+            echo 'export PATH=$ANDROID_SDK_ROOT/cmdline-tools/latest/bin:$PATH'  >> $BASH_ENV
```
Comment (etuleu) [inline .circleci/config.yml:190]: is this change needed here or part of the upgrade emulator and tools?

File: packages/mobile/e2e/src/usecases/HandleDeepLinkSend.js [hunk 8]
```diff
@@ -20,6 +22,7 @@ export default HandleDeepLinkSend = () => {
     // 2. back button
     // there is a slight but important difference because with the back button
     // the activity gets destroyed and listeners go away which can cause subtle bugs
+    await device.reloadReactNative()
```
Comment (etuleu) [inline packages/mobile/e2e/src/usecases/HandleDeepLinkSend.js:25]: do we still need this? or was this the fix that makes it less flaky?

File: packages/mobile/e2e/src/usecases/NewAccountOnboarding.js [hunk 9]
```diff
@@ -25,11 +35,44 @@ export default NewAccountOnboarding = () => {
...
     await element(by.id('VerificationEducationSkipHeader')).tap()
     await element(by.id('VerificationSkipDialog/PrimaryAction')).tap()
 
     // Arrived to Home screen
     await expect(element(by.id('SendOrRequestBar'))).toBeVisible()
   })
+
+  // Ideally this wouldn't be dependent on the previous test
+  // Skip setup on android for now
+  if (device.getPlatform() === 'ios') {
```
Comment (etuleu) [inline packages/mobile/e2e/src/usecases/NewAccountOnboarding.js:51]: yea this is not great if we shuffle the tests around or try to reuse. Why did we need this again?

File: packages/mobile/e2e/src/usecases/ResetAccount.js [hunk 10]
```diff
@@ -1,26 +1,38 @@
-import { enterPinUiIfNecessary, waitForElementId, sleep } from '../utils/utils'
+import { enterPinUiIfNecessary, pixelDiff, waitForElementId } from '../utils/utils'
 import { SAMPLE_BACKUP_KEY } from '../utils/consts'
+import { dismissBanners } from '../utils/banners'
 
 export default ResetAccount = () => {
+  beforeEach(async () => {
+    await device.reloadReactNative()
```
Comment (etuleu) [inline packages/mobile/e2e/src/usecases/ResetAccount.js:7]: do we need this?

File: packages/mobile/e2e/src/usecases/ResetAccount.js [hunk 11]
```diff
@@ -1,26 +1,38 @@
...
+  })
+
   it('Reset Account by doing the Account Key quiz', async () => {
     // This test has very high flakiness on Android. I did my best to fix it, but
     // locally it works every time but on CI it fails 50%+ of the time.
     // TODO: Keep investigating the source of flakiness of this test on Android.
-    if (device.getPlatform() === 'android') {
-      return
-    }
+    // if (device.getPlatform() === 'android') {
```
Comment (etuleu) [inline packages/mobile/e2e/src/usecases/ResetAccount.js:15]: Yes, I think the reset functionality is not working well and we need to look into it.

File: packages/mobile/e2e/src/usecases/RestoreAccountOnboarding.js [hunk 12]
```diff
@@ -1,16 +1,24 @@
...
+import { SAMPLE_BACKUP_KEY, EXAMPLE_NAME } from '../utils/consts'
+import { dismissBanners } from '../utils/banners'
 
 export default RestoreAccountOnboarding = () => {
+  beforeAll(async () => {
```
Comment (etuleu) [inline packages/mobile/e2e/src/usecases/RestoreAccountOnboarding.js:6]: make it into a util function? I think we need it in multiple places.

File: packages/mobile/e2e/src/usecases/SetAccountKey.js [hunk 13]
```diff
@@ -6,7 +6,7 @@ export default SetAccountKey = () => {
     if (device.getPlatform() === 'android') {
       return
     }
-    await element(by.id('Hamburguer')).tap()
+    await element(by.id('Hamburger')).tap()
```
Comment (etuleu) [inline packages/mobile/e2e/src/usecases/SetAccountKey.js:9]: was that the French version of a burguer? :)

File: packages/mobile/scripts/install-android-tools.sh [hunk 14]
```diff
@@ -1,26 +1,31 @@
...
+yes | $SDKMANAGER "platform-tools"
 $SDKMANAGER "platforms;android-29"
 $SDKMANAGER "build-tools;29.0.2"
 $SDKMANAGER "ndk;21.0.6113669"
-$SDKMANAGER "system-images;android-29;default;x86_64"
+yes | $SDKMANAGER "system-images;android-29;default;x86_64"
+# TODO(tom): Pin emulator
 $SDKMANAGER "emulator"
+
+# TODO(tom): Set licenses
```
Comment (etuleu) [inline packages/mobile/scripts/install-android-tools.sh:28]: do we still need to do this?

**Review by etuleu — APPROVED** [review]
Awesome work!

Left a few minor comments but feel free to address in a subsequent PR maybe.
Also maybe edit the title to specify that new tests were added as well as the android command line tools were upgraded.

File: packages/mobile/e2e/src/usecases/RestoreAccountOnboarding.js (hunk 12 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/RestoreAccountOnboarding.js:6]: There is [a util function for restore account onboarding](https://github.com/celo-org/wallet/blob/3138944668ff8699aa99bcd288dde5c2dc307671/packages/mobile/e2e/src/utils/utils.js#L136); however, I keep this restore flow to retain individual test specs the restore flow.

File: packages/mobile/e2e/src/usecases/SetAccountKey.js (hunk 13 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/SetAccountKey.js:9]: Changed all instances of 'Hamburguer' to 'Hamburger' for consistency.

File: .circleci/config.yml (hunk 7 above)
Comment (MuckT) [reply .circleci/config.yml:190]: This change was due to the folder changes with [cmdline tools](https://developer.android.com/studio#downloads) as opposed to the previously used download.

File: packages/mobile/e2e/src/usecases/HandleDeepLinkSend.js (hunk 8 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/HandleDeepLinkSend.js:25]: Currently this test is skipped. More investigation will be needed to determine on what is needed to make it run consistently without disrupting other test specs.

File: packages/mobile/e2e/src/usecases/NewAccountOnboarding.js (hunk 9 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/NewAccountOnboarding.js:51]: This was because currently `.getAttributes()`, needed to parse the account key, isn't working on Android for Detox;  however, a recent update should have enabled it: types just need to be [updated in Detox](https://github.com/wix/Detox/pull/2771).

File: packages/mobile/e2e/src/usecases/ResetAccount.js (hunk 11 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/ResetAccount.js:15]: Agreed we should have an e2e test for this. The current issue is the account shutdown is making it hard for detox to reconnect to the emulator. Hence why we don't actually reset the account and reopen the application.

File: packages/mobile/scripts/install-android-tools.sh (hunk 14 above)
Comment (MuckT) [reply packages/mobile/scripts/install-android-tools.sh:28]: This is to set the licenses based on keys. We are accepting the licenses on the install with the following:
```
yes | $SDKMANAGER "platform-tools"
$SDKMANAGER "platforms;android-29"
$SDKMANAGER "build-tools;29.0.2"
$SDKMANAGER "ndk;21.0.6113669"
yes | $SDKMANAGER "system-images;android-29;default;x86_64"

File: packages/mobile/e2e/src/usecases/ResetAccount.js (hunk 10 above)
Comment (MuckT) [reply packages/mobile/e2e/src/usecases/ResetAccount.js:7]: We do need this with the new test structure so the tests start off in a known state. `await device.reloadReactNative()` will launch the app at the homescreen for a restored account / at the start of education for an unrestored account.
